pkg/bitlpm/trie.go

```diff
@@ -0,0 +1,462 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package bitlpm
+
+// Trie is a non-thread-safe binary trie that indexes arbitrarily long
+// bit-based keys with associated prefixes (that is, "masks") indexed from
+// [most significant bit] ("MSB") to [least significant bit] ("LSB") using
+// the [longest prefix match algorithm].
+//
+// Each method's comments describes the mechanism of how the method
+// works.
+//
+// [most significant bit]: https://en.wikipedia.org/wiki/Bit_numbering#Most_significant_bit
+// [least significant bit]: https://en.wikipedia.org/wiki/Bit_numbering#Least_significant_bit
+// [longest prefix match algorithm]: https://en.wikipedia.org/wiki/Longest_prefix_match
+type Trie[K, T any] interface {
+	// Update updates the trie with a a prefix, key, and value.
+	//
+	// Note:If prefix exceeds the Trie's maximum prefix, it will be set
+	// to the Trie's maximum prefix.
+	Upsert(prefix uint, key K, value T)
+	// Delete removes a key with the given prefix and returns
+	// false if the key was not found.
+	// Note: If prefix exceeds the Trie's maximum prefix, it will be set
+	// to the Trie's maximum prefix.
+	Delete(prefix uint, key K) bool
+	// Lookup lookups any key that matches to at least
+	// the prefix argument.
+	// Note: If prefix exceeds the Trie's maximum prefix, it will be set
+	// to the Trie's maximum prefix.
+	Lookup(prefix uint, key K) T
```

Comment:
Why does lookup require a prefix? Isn't the idea of an LPM lookup that it will perform a lookup with a key and find the most specific prefix that matches?

Reply:
To do a full-match-of-the-key lookup, you specify the full prefix length of the key (if it is a 2-byte key, then 16), but the lookup can also do range lookups (with a less specific prefix).

Reply:
My primary expectation is the other way - Not needing a prefix at all. Seems like for that behaviour I would do `Lookup(0, ...)`. I guess it feels a little odd to me to have to pass this zero value just to get the unqualified longest possible match within the data structure. But then, what's the alternative - have multiple lookup functions tailored for different types of data you're hoping to extract out of the structure? Maybe that's reasonable... throwing ideas out here of different options:

- `Lookup(key K) T` - Find the entry with the longest prefix of `key`
- `LookupRange(key K, prefix uint) []T` or `LookupRange(key K, minPrefix, maxPrefix uint) []T` or (more efficient for allocations) `LookupRange(key K, minPrefix, maxPrefix uint, result []T)` - Find entries within the range of specified prefixes. Seems like it could express the default in the current API while also facilitating the  use case that @squeed brought up below.

But it sounds like you have a particular use case in which you expect there to be some minimum prefix that *must* match in order to perform the lookup, probably related to something in the policy representation - perhaps it would help me to understand if I had a better idea of how you expect to use this API.

Reply:
I've added to lookup methods.